templates/renew-certs.py

```diff
@@ -1,46 +1,97 @@
+    for cert in certs:
+        if os.access(cert['certpath'], os.F_OK):
+            stat = os.stat(cert['certpath'])
+            print 'Certificate file ' + cert['certpath'] + ' already exists'
 
-for cert in certs:
-    if os.access(cert['certpath'], os.F_OK):
-        stat = os.stat(cert['certpath'])
-        print "Certificate file " + cert['certpath'] + " already exists"
+            if time.time() - stat.st_mtime < {{ letsencrypt_min_renewal_age }} * 86400:
+                print ('  The certificate is younger than {{ letsencrypt_min_renewal_age }} days.' +
+                       ' Not creating a new certificate.\n')
+                continue
 
-        if time.time() - stat.st_mtime < {{ letsencrypt_min_renewal_age }} * 86400:
-            print "  The certificate is younger than {{ letsencrypt_min_renewal_age }} days. Not creating a new certificate.\n"
-            continue
+        host = ','.join(cert['host']) if type(cert['host']) is list else cert['host']
 
-    host = ",".join(cert['host']) if type(cert['host']) is list else cert['host']
+        print 'Generating certificate for ' + host
+        args = [
+            '/usr/bin/env', 'python', script,
+            '--account-key',
+            '{{ letsencrypt_account_key }}',
+            '--csr',
+            '{{ acme_tiny_data_directory }}/csrs/' + cert['name'] + '.csr',
+            '--acme-dir',
+            '{{ acme_tiny_challenges_directory }}'
+        ]
 
-    print "Generating certificate for " + host
-    args = [
-        "/usr/bin/env", "python", script,
+        cmd = ' '.join(args)
+        p = Popen(
+            cmd, shell=True, stdin=PIPE, stdout=PIPE, stderr=PIPE, close_fds=True,
+            preexec_fn=demote(user_uid, user_gid), env=env
+        )
+        output = p.stdout.read()
+        p.stdin.close()
+        if p.wait() != 0:
+            print 'error while generating certificate for {}'.format(host)
+            print p.stderr.read()
+        else:
+            f = open(cert['certpath'], 'w')
+            f.write(output)
+            f.close()
+            renewed = True
 
-        "--account-key",
-        "{{ letsencrypt_account_key }}",
-        "--csr",
-        "{{ acme_tiny_data_directory }}/csrs/" + cert['name'] + ".csr",
-        "--acme-dir",
-        "{{ acme_tiny_challenges_directory }}"
-    ]
+    return renewed
 
-    cmd = "/usr/bin/env " + " ".join(args)
 
-    p = Popen(cmd, shell=True, stdin=PIPE, stdout=PIPE, stderr=PIPE, close_fds=True)
-    output = p.stdout.read()
-    p.stdin.close()
-    if p.wait() != 0:
-        print "error while generating certificate for " + host
-        print p.stderr.read()
+def run_post_renew_cmd():
+    """Optionally runs a command after renewal (run as the root user)"""
+    post_renew_cmd = '{{ letsencrypt_post_renew_cmd }}'
+    if not len(post_renew_cmd):
+        print 'No post-renew command configured'
+        return
+
+    process = Popen(
```

Comment:
Yeah, I only used this style, since it was what was used above. It does have the benefit of easily being able to get at the error message, which is something not totally possible with `subprocess.call` (unless you write to/read from a file). `subprocess.check_output` could probably work for these too -- just need to be a little careful about exception handling. I agree that it'd be a bit more concise, but I'm kind of inclined to leave both of them as-is, since this has already been tested with a successful renewal/reload, and it would be a pain to re-test it effectively (letsencrypt doesn't like it when you make too-frequent renewals).